xcdat/coord.py

```diff
@@ -0,0 +1,201 @@
+            )
+
+        try:
+            bounds = self._dataset.cf.get_bounds(coord)
+        except KeyError:
+            bounds = None
+            if bounds is None and not allow_generating:
+                raise ValueError(
+                    f"{coord} bounds were not found in the dataset, bounds must be generated"
+                )
+            else:
+                logger.info(
+                    f"{coord} bounds were not found in the dataset, generating bounds."
+                )
+                bounds = self._generate_bounds(coord, width)
+
+        return bounds
+
+    def _generate_bounds(
+        self,
+        coord: Coord,
+        width: float = 0.5,
+    ) -> xr.DataArray:
+        """Generates bounds for a coordinate using its data points.
+
+        Parameters
+        ----------
+        coord : Coord
+            The coordinate.
+        width : float, optional
+            Width of the bounds relative to the position of the points, by default 0.5.
+
+        Returns
+        -------
+        xr.DataArray
+            The coordinate bounds.
+
+        Raises
+        ------
+        ValueError
+            If coords dimensions does not equal 1.
+        ValueError
+            If coords are length of <=1.
+
+        Notes
+        -----
+        Based on ``iris.coords._guess_bounds`` [1]_ and ``cf_xarray.accessor.add_bounds`` [2]_.
+
+        References
+        ----------
+
+        .. [1] https://scitools-iris.readthedocs.io/en/stable/generated/api/iris/coords.html#iris.coords.AuxCoord.guess_bounds
+
+        .. [2] https://cf-xarray.readthedocs.io/en/latest/generated/xarray.Dataset.cf.add_bounds.html#
+        """
+        da_coord: xr.DataArray = self._get_coord(coord)
+
+        # Validate coordinate shape and dimensions
+        if da_coord.ndim != 1:
+            raise ValueError("Cannot generate bounds for multidimensional coordinates.")
+        if da_coord.shape[0] <= 1:
+            raise ValueError("Cannot generate bounds for a coordinate of length <= 1.")
+
+        # Retrieve coordinate dimension to calculate the diffs between points.
+        dim = da_coord.dims[0]
+        diffs = da_coord.diff(dim)
+
+        # Add a beginning and end point to account for lower and upper bounds.
+        diffs = np.insert(diffs, 0, diffs[0])
+        diffs = np.append(diffs, diffs[-1])
+
+        # Get lower and upper bounds by using the width from nearest point.
+        # Transpose both bound arrays into a 2D array.
+        lower_bounds = da_coord - diffs[:-1] * width
+        upper_bounds = da_coord + diffs[1:] * (1 - width)
+        bounds = np.array([lower_bounds, upper_bounds]).transpose()
+
+        # Clip latitude bounds at (-90, 90)
+        if (
+            da_coord.name in ("lat", "latitude", "grid_latitude")
+            and "degree" in da_coord.attrs["units"]
+        ):
+            if (da_coord >= -90).all() and (da_coord <= 90).all():
+                np.clip(bounds, -90, 90, out=bounds)
```

Comment:
@pochedls @jasonb5 FYI `_generate_bounds()` has been significantly refactored. 

It is no longer based on CDAT's guess_bounds(), which was difficult to understand (specifically the longitude bounds code). Refer to the other comment. I performed an ad-hoc testing on a dataset to validate that the generated bounds equal the original bounds.

Also, it handles generating time bounds and the accessor is now called `CoordAccessor` instead of `AxisAccessor`.